### ISSUE TITLE
How to add channel credentials to client for Authentication with TLS

### ISSUE THREAD

**Issue by George-Madeley** [stephenh/ts-proto#1111]
Hi!

tl;dr How do I add TLS credentials to my gRPC web fetch request in TypeScript?

I have a backend server built using C++ and gRPC. I am now in the process of adding login feature to the backend with JWT. To do this, I have added an interceptor layer which runs an instance of an overridden `AuthMetadataProcessor` class which checks if the desired rpc service requires authentication (i.e., login wont but other services will). However, this wont work using insecure credentials. To solve this, I have added TLS keys and certificates to secure the backend. But now, my frontend (coded in TypeScript and Reactjs), fails to connect to the backend because it lacks the required credentials to perform a handshake with the backend.

I was originally using `proto-gen-js` and `ts-protoc-gen` for my protobuf compiling into JavaScript and TypeScript respectively and proxy to connect to the backend. However, after reading [this DEV article](https://dev.to/notmedia/how-to-setup-and-test-tls-in-grpcgrpc-web-485m#:~:text=There%20are%20two%20implementations%20%E2%80%94%20official#:~:text=There%20are%20two%20implementations%20%E2%80%94%20official) I switched to using `ts-proto`.

In their example, they do the following:

```typescript
import { ChannelCredentials } from '@grpc/grpc-js';
import * as fs from 'fs';
import * as path from 'path';

import { TLSServiceClient } from './generated/proto/tls_service';

function getChannelCredentials(): ChannelCredentials {
  const rootCert = fs.readFileSync(path.resolve(__dirname, '../certs/ca-cert.pem'));
  
  // If you use CA root certificate
  // const channelCredentials = ChannelCredentials.createSsl();

  // If you use Self-Signed root certificate you need to provide it
  const channelCredentials = ChannelCredentials.createSsl(rootCert);

  return channelCredentials;
}

function main() {
  const credentials = getChannelCredentials();

  const client = new TLSServiceClient('0.0.0.0:4000', credentials);

  client.unary({ id: 'test' }, (error, response) => {
    // eslint-disable-next-line no-console
    console.log('response: ', response);
  });
}

main();
```

However, I cannot seem to add credentials to the client constructor like they do in this commaned:

```typescript
const client = new TLSServiceClient('0.0.0.0:4000', credentials);
```

How do I add these credentials to my fetch request?

**Comment by stephenh** [comment]
Hi @George-Madeley , when you say "cannot seem to add credentials", you didn't include why that doesn't work; is it a type error? It blows up at runtime? Etc...

The best I can do is point to our test suite that has an example with the insecure credentials, and that swapping those to be secure credentials should work:

https://github.com/stephenh/ts-proto/blob/ffacf3aaa8fad43c69e75a184c0703fc050b7868/integration/grpc-js/grpc-js-test.ts#L152

Otherwise I can't really help much; if you figure this out, and would like to submit any improvements to our docs/readme to help others, that would be great! Thanks!

**Comment by George-Madeley** [comment]
Hi @stephenh, I figured out my error. I am using gRPC Web connected to a C++ gRPC backend with authentication.

I found an answer for those also stuck on this problem. As far as I am aware, you cannot add TLS authentication credentials between gRPC web client and a gRPC server as mentioned in [this issue](https://github.com/grpc/grpc-web/issues/1179). However, you can add TLS authentication credentials between the gRPC proxy and the gRPC server. This completes the handshake and allows you to use the `AuthMetadataProcessor` class. How to implement TLS on the proxy depends on the proxy you are using. However, the connection is only secure between the proxy and the server; the connection between the client and the proxy remain unsecure.
